catalog/db_branch_reader.go

```diff
@@ -0,0 +1,241 @@
+			if err != nil {
+				panic(err)
+			}
+			r.nextRow[nonNilNextRow[i]] = n
+		}
+	}
+	return selectedEntry, nil
+}
+
+func (r *singleBranchReader) GetNextPK() (*entryPK, error) {
+	if r.EOF {
+		return nil, nil
+	}
+	if r.firstTime {
+		r.firstTime = false
+		r.buf = make([]*entryPK, 0, r.bufSize)
+		q := baseSelect(r.branchID, r.commitID).Limit(uint64(r.bufSize)).Where("path >= ?", r.after)
+		err := fillBuf(r.tx, q, &r.buf)
+		if err != nil {
+			return nil, err // todo: just to trick the LINTER, remove when done
+			// panic(err)
+		}
+	}
+	// returnes the significant entry of that Path, and remove rows with that Path from buf
+	l := len(r.buf)
+	if l == 0 {
+		r.EOF = true
+		return nil, nil
+	}
+	// last Path in buffer may have more rows that were not read yet
+	if *r.buf[l-1].Path == *r.buf[0].Path {
+		err := r.extendBuf()
+		if err != nil {
+			panic(err)
+		}
+		l = len(r.buf)
+	}
+	if l == 0 {
+		r.EOF = true
+		return nil, nil
+	}
+	firstPath := *r.buf[0].Path
+	var i int
+	for i = 1; i < l && *r.buf[i].Path == firstPath; i++ {
+	}
+	nextPK := findSignificantEntry(r.buf[:i], r.commitID)
+	r.buf = r.buf[i:] // discard first rows from buffer
+	return nextPK, nil
+}
+
+func findSignificantEntry(buf []*entryPK, lineageCommitID CommitID) *entryPK {
+	var ret *entryPK
+	l := len(buf)
+	if buf[l-1].MinCommit == 0 { // uncommitted.Will appear only when reading includes uncommitted entries
+		ret = buf[l-1]
+	} else {
+		ret = buf[0]
+	}
+	// if entry was deleted after the max commit that can be read, it must be set to undeleted
+	if lineageCommitID == CommittedID ||
+		lineageCommitID == UncommittedID ||
+		ret.MaxCommit == MaxCommitID {
+		return ret
+	}
+	//todo: rethink condition
+	if ret.MaxCommit >= lineageCommitID {
+		ret.MaxCommit = MaxCommitID
+	}
+	return ret
+}
+
+func baseSelect(branchID int64, commitID CommitID) sq.SelectBuilder {
+	q := sq.Select("branch_id", "path", "min_commit", "max_commit", "ctid").
+		From("catalog_entries").
+		Where("branch_id = ? ", branchID).
+		OrderBy("branch_id", "path", "min_commit desc")
+	if commitID == CommittedID {
+		q = q.Where("min_commit > 0")
+	} else if commitID > 0 {
+		q = q.Where("min_commit between 1 and ?", commitID)
+	}
+	return q
+}
+
+func (r *singleBranchReader) extendBuf() error {
+	l := len(r.buf)
+	if l == 0 {
+		panic("in extendBuf with empty buffer!!!!!")
+	}
+	lastRow := r.buf[l-1]
+	completionQuery := baseSelect(r.branchID, r.commitID)
+	completionQuery = completionQuery.Where("Path = ? and min_commit < ?", *lastRow.Path, lastRow.MinCommit)
+	s := sq.DebugSqlizer(completionQuery)
+	_ = s
+	continueationQuery := baseSelect(r.branchID, r.commitID)
+	continueationQuery = continueationQuery.
+		Where("Path > ?", *lastRow.Path).
+		Limit(uint64(r.bufSize))
+	s = sq.DebugSqlizer(continueationQuery)
+	_ = s
+	// move rows of last Path to beginning of buffer
+	tempBuf := make([]*entryPK, 0, r.bufSize+len(r.buf)*2)
+	tempBuf = append(tempBuf, r.buf...)
```

Comment:
use 'copy'

Reply:
Will change all of this once I change min-commit to MaxCommitID